sds_data_manager/lambda_code/SDSCode/rds_utils/db_ingest_query.py

```diff
@@ -0,0 +1,89 @@
+import logging
+import sys
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+
+class DbIngestQuery:
+    """
+    Class to represent the connection to the ingest query.
+
+    ...
+
+    Attributes
+    ----------
+    query: str
+        the query string created for ingest including placeholders for
+        the ingest data.
+    data: tuple
+        data for the ingest query that will be inserted into the query
+        string by psycopg2.
+    s3_path: str
+        S3 path for the file being ingested.
+    metadata: dict
+        metadata dictionary for the file being ingested.
+
+    Methods
+    -------
+    get_query(query):
+        gets the ingest query string.
+    get_data():
+        gets the ingest data tuple.
+    """
+
+    def __init__(self, s3_path: str, metadata: dict):
+        self.query: str
+        self.data: tuple
+        self.s3_path = s3_path
+        self.metadata = metadata
+
+        self._construct_ingestion_query()
+
+    def _construct_ingestion_query(self):
+        """Constructs the ingestion query string and corresponding data structure."""
+        logger.info("Constructing metadata ingestion query")
+        # add s3 path to the id field
+        self.metadata["id"] = self.s3_path
+        # break up date into year, month, day
+        self.metadata["year"] = self.metadata["date"][:4]
+        self.metadata["month"] = self.metadata["date"][4:6]
+        self.metadata["day"] = self.metadata["date"][6:8]
+        # remove date from metadata
+        self.metadata.pop("date")
+        logger.info(f"metadata: {self.metadata}")
+        # create a string for the fields in query
+        fields = ", ".join(self.metadata.keys())
+        # create placeholders for the query data to put into query string
+        placeholders = ", ".join(["%s"] * len(self.metadata))
```

Comment:
I think I understand the others but not sure of this `placeholders`. Can you give an example of what it's doing with `len(self.metadata)` and what the input to query would look like?

Reply:
Yeah, I will add a comment explaining that. For psycopg you can pass a query string and query data to its execute method to send the query. The query string looks something like this:
```python
"""
INSERT INTO metadata (
    mission, instrument, level, id, year, month, day)
VALUES (%s, %s, %s, %s, %s, %s, %s)
"""
```

So you put the field names in the query string, but leave placeholders (the `%s`s) for the actual data. Then you pass the query string along with a tuple containing the data for each field where each value in the tuple will map to one of the placeholders (`%s`). 

I'm realizing that you might not need to pass in a separate data tuple and could add those values to the string before calling execute, so the placeholders might not be needed. But that's what they're doing here.